internal/app/secretless/providers/conjur/provider.go

```diff
@@ -271,45 +231,40 @@ func (p *Provider) refreshAccessToken() error {
 	// Authenticate in a loop with retries on failure with exponential backoff
 	err = backoff.Retry(func() error {
 		for {
-
 			// Lock the authenticatorMutex
 			p.AuthenticationMutex.Lock()
 
 			log.Printf("Info: Conjur provider is authenticating as %s ...", p.Authenticator.Config.Username)
 			resp, err := p.Authenticator.Authenticate()
-
-			if err == nil {
-				log.Printf("Info: Conjur provider received a valid authentication response")
-				err = p.Authenticator.ParseAuthenticationResponse(resp)
-			}
-
 			if err != nil {
+				p.AuthenticationMutex.Unlock()
 				log.Printf("Info: Conjur provider received an error on authenticate: %s", err.Error())
+				return err
+			}
 
-				if autherr, ok := err.(*authenticator.Error); ok {
-					if autherr.CertExpired() {
-						log.Printf("Info: Conjur certificate expired; Conjur provider is re-logging in.")
-
-						if err = p.Authenticator.Login(); err != nil {
-							return err
-						}
+			log.Printf("Info: Conjur provider received a valid authentication response")
 
-						// if the cert expired and login worked then continue
-						continue
-					}
-				} else {
-					return fmt.Errorf("Error: Conjur provider unable to authenticate: %s", err.Error())
-				}
+			err = p.Authenticator.ParseAuthenticationResponse(resp)
+			if err != nil {
+				p.AuthenticationMutex.Unlock()
+				log.Printf("Failure parsing response: %s", err.Error())
+				return err
 			}
 
-			// Unlock the authenticatorMutex
 			p.AuthenticationMutex.Unlock()
 
+			if oneShot {
+				return nil
+			}
+
 			// Reset exponential backoff
 			expBackoff.Reset()
 
-			log.Printf("Info: Conjur provider is waiting for %v minutes to re-authenticate.", authenticateCycleDuration)
-			time.Sleep(authenticateCycleDuration)
+			log.Printf("Info: Conjur provider is waiting for %s to re-authenticate.",
+				p.Authenticator.Config.TokenRefreshTimeout)
+
+			fmt.Println()
+			time.Sleep(p.Authenticator.Config.TokenRefreshTimeout)
```

Comment:
Is this all still needed if we're sleeping in the goroutine too? Seems redundant

Reply:
👍

Reply:
This is changed by the PR to be a bit more intuitive but yes this is needed here since the main go subroutine (async loop) starts after the initial one is completed in sync manner and we don't want both to run concurrently.